monai/engines/workflow.py

```diff
@@ -121,24 +121,24 @@ def __init__(
         to_kwargs: dict | None = None,
         amp_kwargs: dict | None = None,
     ) -> None:
-        if iteration_update is not None:
-            super().__init__(iteration_update)
-        else:
-            super().__init__(self._iteration)
+        super().__init__(self._iteration if iteration_update is None else iteration_update)
 
         if isinstance(data_loader, DataLoader):
-            sampler = data_loader.__dict__["sampler"]
+            sampler = getattr(data_loader, "sampler", None)
+
+            # set the epoch value for DistributedSampler objects when an epoch starts
             if isinstance(sampler, DistributedSampler):
 
                 @self.on(Events.EPOCH_STARTED)
                 def set_sampler_epoch(engine: Engine) -> None:
                     sampler.set_epoch(engine.state.epoch)
 
+            # if the epoch_length isn't given, attempt to get it from the length of the data loader
             if epoch_length is None:
-                epoch_length = len(data_loader)
-        else:
-            if epoch_length is None:
-                raise ValueError("If data_loader is not PyTorch DataLoader, must specify the epoch_length.")
+                try:
+                    epoch_length = len(data_loader)
+                except TypeError:  # raised when data_loader is given an iterable dataset which has no length
```

Comment:
Do you think we should add a warning here or validate the dataset instance in the `data_loader`?
```
isinstance(data_loader.dataset, Iterable)
```

Reply:
The `data_loader` argument itself isn't always a `DataLoader` instance so may not have a `dataset` member, it can be a `Iterable` itself which might have a length. It think there's a use case where you'd pass in just a list of items you already have a so not use a `DataLoader` at all and so this is the more robust way of attempting to get a length.

Reply:
So it looks like `if epoch_length is None:` this if should be moved to the outside level, right?

Reply:
I've tried a different method here actually that avoid checking the exception throw.

Reply:
And then I reverted to something similar again.